pkg/arvo/sys/hoon.hoon

```diff
@@ -1897,6 +1938,483 @@
+    ==
+  ::
+  ++  lldub
+    |*  a=(lnode)
+    ^-  (ltree _k.n.a _v.n.a)
+    ?>  ?=(^ r.a)
+    ?-  -.r.a
+        %llos  (llsin n.a l.a m.a (lrsin p.r.a))
+        %rlos  (llsin n.a l.a m.a (rrsin p.r.a))
+    ==
+  ::
+  ++  lrdub
+    |*  a=(lnode)
+    ^-  (ltree _k.n.a _v.n.a)
+    ?>  ?=(^ l.a)
+    ?-  -.l.a
+        %llos  (lrsin n.a (llsin p.l.a) m.a r.a)
+        %rlos  (lrsin n.a (rlsin p.l.a) m.a r.a)
+    ==
+  ::
+  ++  rldub
+    |*  a=(lnode)
+    ^-  (ltree _k.n.a _v.n.a)
+    ?>  ?=(^ r.a)
+    ?-  -.r.a
+        %llos  (rlsin n.a l.a m.a (lrsin p.r.a))
+        %rlos  (rlsin n.a l.a m.a (rrsin p.r.a))
+    ==
+  ::
+  ++  rrdub
+    |*  a=(lnode)
+    ^-  (ltree _k.n.a _v.n.a)
+    ?>  ?=(^ l.a)
+    ?-  -.l.a
+        %llos  (rrsin n.a (llsin p.l.a) m.a r.a)
+        %rlos  (rrsin n.a (rlsin p.l.a) m.a r.a)
+    ==
+  ::
+  ++  toy                                               ::  play
+    |*  [a=(pri) b=(pri)]
+    ^+  a
+    ?~  a
+      b
+    ?~  b
+      a
+    ?:  (win [p.n.a k.n.a] [p.n.b k.n.b])
+      [n.a (rbal n.b t.a m.a t.b) m.b]
+    [n.b (lbal n.a t.a m.a t.b) m.b]
+  ::
+  ++  sec                                               ::  second best
+    |*  [t=(ltree) m=*]
+    |-
+    ^-  (pri _?>(?=(^ t) k.n.p.t) _?>(?=(^ t) v.n.p.t))
+    ?~  t  ~
+    ?-  -.t
+      %llos  (toy [n.p.t l.p.t m.p.t] $(t r.p.t))
+      %rlos  (toy $(t l.p.t, m m.p.t) [n.p.t r.p.t m])
+    ==
+  ::
+  ++  win                                               ::  compare
+    |*  [[p=@ q=*] [r=@ s=*]]
+    ?|  (lth p r)
+        ?&(=(p r) (gor q s))
+    ==
+  --
+  ::
+  =|  a=(pri)
+  |@
+  ++  deb  +>                                           ::  internals access
+  ::
+  ++  wyt  ?~(a 0 +((size t.a)))                        ::  queue size
+  ::
+  ++  get                                               ::  get binding by key
+    |*  k=*
+    =>  .(k `_?>(?=(^ a) k.n.a)`k)
+    |-  ^-  (unit (pair @ _?>(?=(^ a) v.n.a)))
+    ?~  a  ~
+    =/  tor=(torn _?>(?=(^ a) k.n.a) _?>(?=(^ a) v.n.a))
+      see
+    ?~  tor  ~
+    ?-    -.tor
+        %sing
+      ?.  =(k k.n.tor)
+        ~
+      (some [p.n.tor v.n.tor])
+    ::
+        %play
+      ?:  (gor k top(a l.tor))
+        $(a l.tor)
+      $(a r.tor)
+    ==
+  ::
+  ++  has                                               ::  key existence check
+    |*  k=*
+    !=(~ (get k))
+  ::
+  ++  min                                               ::  lowest-pri element
+    ^-  (unit _?>(?=(^ a) n.a))
+    ?~  a
+      ~
+    (some n.a)
+  ::
+  ++  put                                               :: add [key prio val]
+    |*  [k=* p=@ v=*]
+    |-  ^+  a
+    ?~  a
+      (one k p v)
+    ?:  ?=(~ t.a)
+      ?:  =(k m.a)
+        (one k p v)
+      ?:  (gor k m.a)
+        (toy (one k p v) (one n.a))
+      (toy (one n.a) (one k p v))
+    ?-    -.t.a
+        %rlos
+      ?:  (gor k m.p.t.a)
+        (toy $(a [n.a l.p.t.a m.p.t.a]) [n.p.t.a r.p.t.a m.a])
```

Comment:
Maybe just me, but these long `toy` statements seem like they'd be a lot clearer in tall-form.

Reply:
I think these are ok here.  I feel like this code, with all its gory balancing internals and such, is necessarily pretty terse anyway, so making these expressions tall would sort of stretch things out without adding much real clarity.

I'll mark this as resolved for now, but happy to hear additional/other opinions along these lines.